### ISSUE TITLE
Missing overflow check for AccessList cost addition in IntrinsicGas function

### ISSUE THREAD

**Issue by Kourin1996** [ethereum/go-ethereum#30880]
Geth version: `1.14.13-unstable`
Commit hash : `master` branch (`08e6bdb550712503873fb2a138b30132cc36c481`)

`IntrinsicGas` in `core/state_transition.go` calculates intrinsic gas for a given transaction. Cost calculation of `data` field take care of overflow while the calculation for Access List doesn't. (Please see code below)

https://github.com/ethereum/go-ethereum/blob/08e6bdb550712503873fb2a138b30132cc36c481/core/state_transition.go#L104-L115

In general, a transaction causing overflow at this point might be blocked by another check (e.g. TxPool slot limit). But `evm transaction` command which validates a transaction offline might return wrong result by this problem.

Ideally, `IntrinsicGas` should check overflow in calculation for AccessList cost like this:

```go
if accessList != nil {
	numItems := uint64(len(accessList))
	if (math.MaxUint64-gas)/params.TxAccessListAddressGas < numItems {
		return 0, ErrGasUintOverflow
	}
	gas += numItems * params.TxAccessListAddressGas

	numStorageKeys := uint64(accessList.StorageKeys())
	if (math.MaxUint64-gas)/params.TxAccessListStorageKeyGas < numStorageKeys {
		return 0, ErrGasUintOverflow
	}
	gas += numStorageKeys * params.TxAccessListStorageKeyGas
}
```

**Comment by holiman** [comment]
Well.. for that to overflow, the length of the accesslist would have to be somewhere on the order of `7e15`? 
```
>>> 0xffffffffffffffff // 2400
7686143364045646
```
Might be that the preceding calculation on `data` is overly paranoid: I don't see a reason to make the accessList implementation go to the same lengths. 

?

**Comment by rjl493456442** [comment]
Yeah, but to be fair, the `TxDataZeroGas` overflow checking is also theoretical needed but in practice unnecessary.

I would lean to align the behaviors.

**Comment by MariusVanDerWijden** [comment]
> We have made up our mind, and do not want to reduce the readability of the code for a case that cannot happen.
